### ISSUE TITLE
Investigate difference between `torch.eig` and `numpy.linalg.eig`

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#4533]
Reported here: https://discuss.pytorch.org/t/torch-eig-seems-really-unstable-can-anyone-explain-this-result/11579

Users may assume that `torch.eig` will return orthornomal eigenvectors when the input is symmetric. This is the case for `torch.symeig` but I don't think `torch.eig` has this guarantee, so I'll probably add a warning to the docs.

However, for most symmetric matrix inputs, it looks like `torch.eig` does give orthonormal eigenvectors. I'm not sure if the underlying implementation is supposed to do this for all symmetric inputs or not.

The interesting thing is that for some symmetric inputs,`torch.eig` and `numpy.linalg.eig` will give different results that might warrant investigation. For example, for the following input:

```
import numpy as np
L = np.array([[  9.99999881e-01,  -9.82382353e-09],
       [ -9.82382353e-09,   9.99999940e-01]], dtype=np.float32)

numpy.linalg.eig(L)[1] 
torch.eig(torch.Tensor(L), eigenvectors=True)[1]
```

`numpy.linalg.eig` returns orthonormal eigenvectors while `torch.eig` doesn't.

**Comment by t-vi** [comment]
I think there are even more fundamental differences: When `torch.eig` encounters a matrix with complex eigenvalues, it will return a `size([0])`-Tensor and useless (as far as I can tell) eigenvectors while numpy returns a complex result.

Purely imaginary:
```
a = torch.tensor([[0.0,1],[-1,0]])
a.eig()
```
gives
```
(tensor([[ 0.,  1.],
         [ 0., -1.]]), tensor([]))
```
and similarly in the mixed real and imaginary case
```
a2 = torch.eye(4)
a2[:2,:2] = a
a2.eig()
```
gives
```
(tensor([[ 0.,  1.],
         [ 0., -1.],
         [ 1.,  0.],
         [ 1.,  0.]]), tensor([]))
```
I think we might want to prune the eigenvectors as well?

**Comment by t-vi** [comment]
An additional difference seems to be that torch.eig does returns only the real part of complex eigenvectors. (Apologies for the inaccurate previous mail.)
```
a = torch.tensor([[0.0,1],[-1,0]])
a.eig(eigenvectors=True)
```
gives
```
(tensor([[ 0.,  1.],
         [ 0., -1.]]), tensor([[ 0.7071,  0.0000],
         [ 0.0000,  0.7071]]))
```
while `numpy.linalg.eig(a.numpy())` gives
```
(array([ 0.+1.j,  0.-1.j], dtype=complex64),
 array([[ 0.70710677+0.j        ,  0.70710677-0.j        ],
        [ 0.00000000+0.70710677j,  0.00000000-0.70710677j]], dtype=complex64))
```
The PyTorch version seems to be similar to what Lapack / Magma implement in geev_s/d (though magma also has versions doing complex eigenvectors).

**Comment by ailzhang** [comment]
This has been resolved by the doc update #10315 I think. Please reopen if necessary.
